### PR TITLE
Rust SDK http handler macro improvements

### LINKED ISSUE #1372: Improve Rust SDK HTTP handler macro with an `IntoResponse` trait
Right now the http handler in the Rust SDK assumes that the annotated function will return `Result<T, E> where T: TryInto<spin_http::Response>` and where `E` has the same API as `std::error::Error` (which one should note `anyhow::Error` does not implement but has a similar enough API that things just work). This can make error handling in spin http less than ideal. If the user wants to return a non-500 error they need to construct a response using something like `http::ResponseBuilder` which has an awkward API. 

## Proposal 

Instead, we could take some inspiration from other HTTP frameworks and make the response handling a bit more fluid. My proposal would be that we could require that the handler return anything that implements an `IntoResponse` trait that ultimately gets used to turn the user's return value into a response. 

Ultimately, this could end up with something that looks like this:

```rust
fn handler(req: Request) -> anyhow::Result<impl IntoResponse> {
    let query = req.uri().query();
    let Some(query) = query else { return Ok((StatusCode::BAD_REQUEST, "no query string")) };
    let Ok(query) = serde_qs::from_str::<Query>(query) else {
        return Ok((StatusCode::BAD_REQUEST, "could not parse query string"));
    };

    let store = key_value::Store::open_default()?;
    let mut filter = get_state(&store)?;

    let status = match filter.exists(&query.email) {
        Exists::Maybe if expensive_user_lookup(&query.email)? => 409,
        Exists::No | Exists::Maybe => 200,
    };

    Ok((StatusCode::from_u16(status).unwrap(), ""))
}
```

`IntoResponse` could be extended for any types that look like a response.

## Issues 

The only issue is that making this work in a backwards compatible way will be difficult. This is largely because the macro currently assumes a `Result` type where the error is anything that kind of looks like a `std::error::Error`. Importantly it does not require that that error type actually *implement* `std::error::Error` (e.g., just as `anyhow::Error` does not). 

Due to limitations in the Rust type system, it's basically impossible to do the following:

```rust
impl IntoError for anyhow::Error {
  // ...
}

impl <E: std::error::Error> IntoError for E {
  // ...
}
```

This does not work because the Rust is conservatively allowing future versions of `anyhow` to implement `std::error::Error` for `anyhow::Error`. 

## Next steps

So with that in mind, the next questions are:
* Is such an API change worth it?
* If so, can anyone think of a clever way to make this a backwards compatible change?
* If not, how do we introduce a second macro in a way that makes sense?

### PR DESCRIPTION
**Pull request by rylev**
Fixes #1372 

## What are we trying to accomplish?

There are several goals with this change:

* Remove requiring that users use the `http` and `bytes` crates. These crates (especially `bytes`) can be hard to use. Users unfamiliar with them might be annoyed to have to learn about two additional dependencies before being able to use these handlers. At the same time, `http` and `bytes` are common crates, and for those users who want to use the crates, we should integrate seemlessly with them. 
* Allow more response body types: Lots of users want to be able to return String or json objects as their response object. We should provide an easy way to return a rich variety of response body types. 
* Allow convenient ways for users to access their requests. For example, don't require users to manually parse the request body as JSON - give a way for the macro to do that for the user. 
* Allow usage of the same API for `wasi_http` handlers as `http` handlers. Only when the user really needs streaming response bodies should they be required to switch to the more advanced native `wasi-http` API.

## The changes
* The http handler no longer relies on the user of the SDK bringing any crate but `spin_sdk` into scope. This means that the `http_component` macro can be used without the `http` or `bytes` crates being available. 
* The function annotated with `http_component` must have the following signature
```rust 
fn<Request, Response, Error>(req: Request) -> Response 
  where 
    Request: spin_sdk::http::TryFromRequest, 
    Response: spin_sdk::http::IntoResponse, 
```
* The above means that handler functions can be pretty flexible. For example, the following works:
```rust
#[http_component]
fn handle_request(_req: http::Request<Option<String>>) -> anyhow::Result<impl IntoResponse> {
  Ok((200, "Hello, world!"))
}
```
* Note above that the request type uses the `http` crate, but it doesn't have to. `spin_sdk::http::Request` would also work for instance. Also, the return type is a `anyhow::Result<(u16, &str)>` which works because `Result<(u16, &str), anyhow::Error>` implements `IntoResponse`.

### Breaking Changes

I believe the only breaking change is that `spin_sdk::http::{Request, Response}` are no longer aliases to the `http` crate's `Request<Option<bytes::Bytes>>` and `Response<Option<bytes::Bytes>>`. Instead they point to `wit` `http` interface Request and Response types.

### CODE REVIEW

**Comment by ThorstenHans** [comment]
I love this proposal @rylev, 

Especially returning `Result<impl IntoResponse>` is nice, aligns with idioms established by several popular web frameworks in the Rust ecosystem,  and comes pretty handy (thinking of individual trait implementations). 

However, I would not remove `Result` from the return type. Especially not in combination with `anyhow`. It allows the app developer to return early using the `?` operator. I guess that devs must write more boilerplate code when changing the return type to become just `impl IntoResponse`, or do you have something in mind that I don't get at this point? 😄 

IMO, removing `http` and `bytes` would be nice. That said, I use `http` in every app to eliminate magic strings and numbers by using consts like `http::StatusCode::OK` and others.... I would welcome this modification as long as the `spin_sdk` provides structs and consts like `http:StatusCode` (and others, obviously). 

What do you think about `anyhow`? I prefer keeping that one as a dependency. It makes life so much easier and streamlines writing apps a lot.

**Comment by rylev** [comment]
@ThorstenHans thanks for the feedback! So my plan (which seems to be working) is to support the use of `anyhow` and `http` without *requiring* that they are used. So if you find `http` and `anyhow` to be helpful you can use them and everything will just work™, but if you don't want to use them you do need to. As an example, a "hello, world" style handler can be written in all of these ways:

With `anyhow` and `http`:
```rust
#[http_component]
fn handle_request(_req: http::Request<()>) -> anyhow::Result<impl IntoResponse> {
  Ok((200, "Hello, world!"))
}
```

Without `http`:

```rust
#[http_component]
fn handle_request(_req: spin_sdk::http::Request) -> anyhow::Result<impl IntoResponse> {
  Ok((200, "Hello, world!"))
}
```

Without `anyhow`:

```rust
#[http_component]
fn handle_request(_req: spin_sdk::http::Request) -> impl IntoResponse {
  (200, "Hello, world!")
}
```

We'll need to decide if using `anyhow` and `http` is the right thing to show in docs and as part of the templates (perhaps it is), but the new handler is really flexible in what is accepts. For example, the request type is also flexible. This also works:

```rust
use spin_sdk::http::{IntoResponse, Json};

#[derive(serde::Deserialize, Debug)]
struct MyBody {
  data: String
}

#[http_component]
fn handle_request(Json(body): Json<MyBody>) -> impl IntoResponse {
  println!("Received body: {body:?}");
  (200, "Hello, world!")
}
```

**Comment by ThorstenHans** [comment]
@rylev 

You're right! Having the flexibility is great and definitely worth introducing the breaking change you mentioned. However, my concerns were more about documentation and templates. 

Showcasing and teaching Spin with `anyhow` lets developers grow quickly from simply returning `(200, "Hello World")`, to something more realistic that may require key-value store or other surroundings.

Obviously, one could easily modify the signature of the `fn` at any time and bring in whatever is needed...

Bottom line, I would say that `anyhow` makes the code streamlined, short, and left-aligned 🥳, although it's definitely not necessary for every scenario!

File: sdk/rust/src/http/router.rs [hunk 1]
```diff
@@ -103,79 +105,96 @@ impl Router {
...
+            .unwrap();
     }
 
     /// Register a handler at the path for the specified HTTP method.
-    pub fn add<F>(&mut self, path: &str, method: http_types::Method, handler: F)
+    pub fn add<F, I, O>(&mut self, path: &str, method: Method, handler: F)
     where
-        F: Fn(Request, Params) -> Result<Response> + 'static,
+        F: Fn(I, Params) -> O + 'static,
+        I: TryFrom<Request>,
```
[the remark this answers is not in the document]
Comment (rylev) [reply sdk/rust/src/http/router.rs:126]: I think we probably want to make a `FromRequest` trait. I'll be honest, that at this point I'm slowly moving toward 100% custom conversion traits not because we need them right now, but because the orphan rule scares me.

**Review by itowlson — COMMENTED** [review]
As you suggested, I focused on the examples as being most indicative of the user experience.  I love the flexibility, extensibility and expressiveness of the new design, and there are some areas that are perpetual pain points for me where this makes it _way_ simpler.  However, there is certainly more magic involved, and that makes some examples non-obvious how to extend or how they behave.

I appreciate that Rust veterans will eat this stuff for breakfast, and I like having this great offering for power users.  I'm concerned, though, that a good number of people kicking the tyres of Wasm and Spin will be using Rust not because they are veterans but because it has better Wasm support than their language of choice, and they may find the different idioms and sophisticated types a bit intimidating.  Is there a way to offer them simplicity and regularity without compromising the new capabilities?

File: examples/rust-outbound-mysql/src/lib.rs [hunk 2]
```diff
@@ -117,12 +106,10 @@ fn list() -> Result<Response> {
...
     );
 
-    Ok(http::Response::builder()
-        .status(200)
-        .body(Some(response.into()))?)
+    Ok(http::Response::builder().status(200).body(Some(response))?)
 }
 
-fn get(id: i32) -> Result<Response> {
+fn get(id: i32) -> Result<Response<Option<String>>> {
```
[the remark this answers is not in the document]
Comment (rylev) [reply examples/rust-outbound-mysql/src/lib.rs:112]: I actually hasn't - the previous `Result<Response>` was just an alias for `Result<Response<Option<bytes::Bytes>>, Error>`. I personally think that the hiding the body type behind an alias makes this harder to understand and despite the verbosity, it's clearer what this code expects. That being said, this is just an example, so if others feel strongly that hiding the body type behind an alias is the right approach, we can add a `type Response = http::Response<Option<String>>` to the top of the file. I personally think that that is an example of something being more terse but ultimately harder to understand.

File: examples/rust-outbound-mysql/src/lib.rs [hunk 3]
```diff
@@ -43,21 +45,17 @@ fn parse_request(req: Request) -> RequestAction {
...
-                    let is_finicky = map
-                        .get("is_finicky")
-                        .map(|s| s == "true")
-                        .unwrap_or_default();
-                    RequestAction::Create(name, prey, is_finicky)
-                }
-                Err(_) => RequestAction::Error(400), // Sorry no this isn't helpful either
-            }
+            let map = req.body();
+            let name = match map.get("name") {
```
[the remark this answers is not in the document]
Comment (rylev) [reply examples/rust-outbound-mysql/src/lib.rs:49]: There's no reason - the example worked this way before so I didn't change it. Happy to use a concrete struct type instead.

File: examples/http-rust/src/lib.rs [hunk 4]
```diff
@@ -1,15 +1,14 @@
...
 
 /// A simple Spin HTTP component.
 #[http_component]
-fn hello_world(req: Request) -> Result<Response> {
-    println!("{:?}", req.headers());
-    Ok(http::Response::builder()
-        .status(200)
-        .header("foo", "bar")
-        .body(Some("Hello, Fermyon!\n".into()))?)
+fn hello_world(Json(body): Json<MyBody>) -> anyhow::Result<impl IntoResponse> {
```
[the remark this answers is not in the document]
Comment (rylev) [reply examples/http-rust/src/lib.rs:11]: If the argument types to the handler function can't be extracted from the request (using the `TryFromRequest` trait, the error returned from the conversion (which must always implement `IntoResponse`) gets turned into a response and returned. 

In this particular example, the `Json` type has an `TryFromRequest` implementation that returns a `JsonBodyError` which implements `IntoResponse` ultimately returning a "400 - BAD REQUEST" with a nice error message.

File: examples/http-rust-router-macro/src/lib.rs [hunk 5]
```diff
@@ -22,12 +20,9 @@ mod api {
...
     // /hello/:planet
-    pub fn hello_planet(_req: Request, params: Params) -> anyhow::Result<Response> {
+    pub fn hello_planet(_req: Request, params: Params) -> anyhow::Result<impl IntoResponse> {
         let planet = params.get("planet").expect("PLANET");
 
-        Ok(http::Response::builder()
-            .status(http::StatusCode::OK)
-            .body(Some(planet.to_string().into()))
-            .unwrap())
+        Ok((200, planet.to_string()))
```
[the remark this answers is not in the document]
Comment (rylev) [reply examples/http-rust-router-macro/src/lib.rs:26]: This is lifted directly from the popular `axum` library. If we want to limit the number of "helpful but perhaps more advanced" types that implement `IntoResponse` we can do that and slowly introduce them as we gain more experience with this API.

File: examples/http-rust/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (itowlson) [reply examples/http-rust/src/lib.rs:11]: Nice!

File: examples/spin-wagi-http/http-rust/src/lib.rs [hunk 6]
```diff
@@ -1,15 +1,12 @@
...
-use spin_sdk::{
-    http::{Request, Response},
-    http_component,
-};
+use spin_sdk::http_component;
 
 /// A simple Spin HTTP component.
 #[http_component]
-fn goodbye_world(req: Request) -> Result<Response> {
+fn goodbye_world(req: http::Request<()>) -> Result<http::Response<&'static str>> {
```
[the remark this answers is not in the document]
Comment (rylev) [reply examples/spin-wagi-http/http-rust/src/lib.rs:6]: Again, that's because the previous example was an alias for `Request<Option<bytes::Bytes>>` so while it looked simpler at first glance, I'd argue it wasn't (especially considering the `bytes::Bytes` type is actually really user unfriendly IMO. We can offer type aliases if we think that that would be helpful, but I want to avoid falling into the trap of thinking that this diff which superficially sees the edition of more characters is necessarily more complicated when I'd argue it's not.

**Comment by rylev** [comment]
@itowlson the previous way these handlers worked continues to work so this is really only an extension to what the user can write rather than a change. 

The only reason this is a breaking change is that the SDK used to give preferential treatment to the `http` and `bytes` crates by hardcoding the exported `Request` and `Response` types to be aliases to `http::Request<Option<bytes::Bytes>>` and `http::Response<Option<bytes::Bytes>>`. I believe this was a mistake. While it was certainly less characters to type, how to actually use these `Request` and `Response` types was not that straight forward.

For example, previously the user saw the handler's return type was `Result<Response>`. Looks simple enough, but what is this `Response` type? The user then needs to understand that `Response` (which syntactically is an export of the `spin_sdk`) is actually just an alias to the `http` crate's `Response` type. What's more, this `Response` type is generic and the `Response` alias hard codes the generic to `Option<bytes::Bytes>`. They need to then understand how the `bytes` crate works. 

This API was forced on the user, so not only did the examples show that this is how the API worked - it was the only way the API *could* work. Now, we have the flexibility to still support that use case if a user is familiar with `http` and `bytes`, but we can offer much simpler examples should we choose to. 

Instead of `Result<Response>` we can just have examples that return `String` or `anyhow::Result<String>` or `(u16, String)` or the actual underlying `Response` type that is defined in wit instead of a complicated (albeit powerful) type from a third party `http` crate. 

I'll admit, I'm not 100% sure how exactly we should position our examples. That's a tough question that we perhaps won't be able to answer fully until we get actual user feedback, but this is no longer an API design choice and now just a choice of how we want to write examples and docs.

File: sdk/rust/src/http.rs
Comment (lann) [inline sdk/rust/src/http.rs:1]: Should probably split this file up...

File: examples/rust-outbound-mysql/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (radu-matei) [reply examples/rust-outbound-mysql/src/lib.rs:112]: This is really the crux of this set of changes.

In the very first version of the Rust SDK, we prioritized a really simple getting started experience that would hide a lot of the complexity around the types used for the request and response bodies.

This worked great when doing simple things (and for beginner users), however it was really inflexible for a lot of scenarios beyond simple examples.

While seeing `Result<Response>` _might_ seem easier, hiding that only to expose the user to the real (and *inflexible*) return type that is actually `Result<Response<Option<Bytes>>>` is arguably far more difficult to grasp, and simply inflexible.

Finally, this PR brings writing HTTP handlers for Spin in line with what you would expect when writing a handler in other web libraries in Rust.
We have always talked about the audience for a Spin SDK -- developers familiar with the language. This change moves in that direction.

File: examples/spin-wagi-http/http-rust/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (radu-matei) [reply examples/spin-wagi-http/http-rust/src/lib.rs:6]: As the person who initially made the change to have bodies be `Option<Bytes>`, that has proven to be really inflexible and not in line with what a Rust developer would expect when writing a handler in a web context.

We can consider having aliases for special (maybe common?) scenarios like request bodies.

Finally, perhaps it would be easier for the example just to return a `Response<String>` here?

**Review by lann — APPROVED** [review]

File: examples/http-rust-outbound-http/outbound-http/src/lib.rs [hunk 7]
```diff
@@ -1,17 +1,17 @@
...
-    http::{Request, Response},
+    http::{IntoResponse, Request},
     http_component,
 };
 
 /// Send an HTTP request and return the response.
 #[http_component]
-fn send_outbound(_req: Request) -> Result<Response> {
-    let mut res = spin_sdk::outbound_http::send_request(
+fn send_outbound(_req: Request) -> Result<impl IntoResponse> {
```
Comment (michelleN) [inline examples/http-rust-outbound-http/outbound-http/src/lib.rs:9]: running into `destination-not-allowed (error 1)` triggering this component.

Comment (michelleN) [reply examples/http-rust-outbound-http/outbound-http/src/lib.rs:9]: I don't get why that would be happening given the changes are unrelated but looking into it.

**Review by dicej — APPROVED** [review]
Looks like a step forward overall -- thanks for doing this.  I like the more explicit types and the flexibility provided by the conversion utilities.

Please see my comment about error handling in the `wasi-http` streaming example.  There's no obvious right answer, and each option has its own tradeoffs, but I feel that returning `Result` from the incoming request handler can help ensure errors are logged consistently without unduly burdening the app developer.

**Review by fibonacci1729 — APPROVED** [review]

File: examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs [hunk 8]
```diff
@@ -1,34 +1,32 @@
...
+use spin_sdk::wasi_http_component;
+use url::Url;
 
 const MAX_CONCURRENCY: usize = 16;
 
 #[wasi_http_component]
-async fn handle_request(request: IncomingRequest, response_out: ResponseOutparam) -> Result<()> {
-    let method = request.method();
-    let path = request.path_with_query();
+async fn handle_request(request: IncomingRequest, response_out: ResponseOutparam) {
```
[the remark this answers is not in the document]
Comment (rylev) [reply examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs:14]: My motivation was the worry that users will be confused when they use `?` and no response gets returned. I'm worried that this is a case where `?` is too convenient and you end up with confused users not sure why they're not seeing responses. Perhaps though the fact that we're logging the error consistently is good enough?

File: examples/http-rust-outbound-http/outbound-http/src/lib.rs (hunk 7 above)
Comment (rylev) [reply examples/http-rust-outbound-http/outbound-http/src/lib.rs:9]: Good catch. Here's the fix: https://github.com/fermyon/spin/pull/1887/commits/e7d8a175d9a5cd2233c46ab0a66da9829ef14111

File: examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (dicej) [reply examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs:14]: Yeah, I agree that's a hazard.  If we wanted, the SDK could pass its own version of `ResponseOutparam` to the handler function that wraps the real one in an `Rc<RefCell<Option<_>>>`, keeping a reference of its own.  Then, if the handler returns without sending a response (maybe with an error, or maybe it just forgot to call `ResponseOutparam.set` for some reason), the SDK can use its reference to call `ResponseOutparam.set` with a 500 status.

File: examples/http-rust-router-macro/src/lib.rs [hunk 9]
```diff
@@ -1,18 +1,16 @@
...
     let router = http_router! {
         GET "/hello/:planet" => api::hello_planet,
-        _   "/*"             => |_req, params| {
+        _   "/*"             => |_req: Request, params| {
             let capture = params.wildcard().unwrap_or_default();
-            Ok(http::Response::builder()
-                .status(http::StatusCode::OK)
-                .body(Some(capture.to_string().into()))
-                .unwrap())
+            Response::new(200, capture.to_string())
```
Comment (michelleN) [inline examples/http-rust-router-macro/src/lib.rs:13]: I'm not seeing the body as output when triggering this component.

Comment (rylev) [reply examples/http-rust-router-macro/src/lib.rs:13]: Hmmm... It seems to be working for me:

```
$ curl -i localhost:3000/wow       
HTTP/1.1 200 OK
content-length: 3
date: Wed, 18 Oct 2023 07:03:05 GMT

wow
```

File: examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (rylev) [reply examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs:14]: That could work though it doesn't help if the user sets the `ResponseOutParam` and then after returns an error. Not having a return type at least indicates to the user that they are responsible for handling errors and there can't be a misguided assumption that Spin will do it for them. 

Should we start with no return type and if we see in practice that this causes lots of issues, move towards adding one? I believe we should be able to do this in a backwards compatible way.

File: examples/spin-wagi-http/http-rust/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (rylev) [reply examples/spin-wagi-http/http-rust/src/lib.rs:6]: Going to merge this PR first, and then we can take another closer look at the exact set of examples in another PR.

**Comment by rylev** [comment]
Merging this now as the feedback has settled into smaller issues that need additional discussion and can be solved in follow up PRs. Such feedback is probably better suited to smaller PRs where we can have more focused discussion.
